### PR TITLE
Moving examples over to execute transaction

### PR DESCRIPTION
**Pull request by donnfelker**
Updated all the examples to use `executeTransaction`. Also updated the comments in the Realm javadocs to use `executeTransaction`. There are a few `executeTransaction` calls that remain in the examples: 

```
Targets
    Occurrences of '.beginTransaction' in Project
Found Occurrences  (3 usages found)
    Unclassified occurrence  (3 usages found)
        io.realm.examples.appmodules  (1 usage found)
            ModulesExampleActivity.java  (1 usage found)
                onCreate(Bundle)  (1 usage found)
                    ModulesExampleActivity.java  (1 usage found)
                        (135: 21) defaultRealm.beginTransaction();
        io.realm.examples.json  (1 usage found)
            JsonExampleActivity.java  (1 usage found)
                loadJsonFromStream()  (1 usage found)
                    JsonExampleActivity.java  (1 usage found)
                        (100: 14) realm.beginTransaction();
        io.realm.examples.realmgridview  (1 usage found)
            GridViewExampleActivity.java  (1 usage found)
                loadCities()  (1 usage found)
                    GridViewExampleActivity.java  (1 usage found)
                        (126: 14) realm.beginTransaction();
```

These are all special cases, therefore I left them using the manual `begin`, `commit` and `cancel` methods.

### CODE REVIEW

**Comment by donnfelker** [comment]
This is for #2219

**Comment by zaki50** [comment]
@donnfelker `:unitTestExample:testDebugUnitTest` task in `example` seems to fail.
Can you check that?

**Comment by donnfelker** [comment]
@zaki50  Yeah I will check it out ASAP

File: examples/kotlinExample/build.gradle [hunk 1]
```diff
@@ -1,7 +1,8 @@
 buildscript {
-    ext.kotlin_version = '1.0.0-rc-1036'
+    ext.kotlin_version = '1.0.0'
     repositories {
         jcenter()
+        mavenCentral()
```
Comment (cmelchior) [inline examples/kotlinExample/build.gradle]: Any reason to add this?

**Comment by cmelchior** [comment]
```
io.realm.examples.unittesting.ExampleActivityTest > shouldBeAbleToAccessActivityAndVerifyRealmInteractions FAILED
    org.mockito.exceptions.verification.TooLittleActualInvocations at ExampleActivityTest.java:195

io.realm.examples.unittesting.ExampleRealmTest > shouldVerifyTransactionWasCreated FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at ExampleRealmTest.java:106

5 tests completed, 2 failed
:unitTestExample:testDebugUnitTest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':unitTestExample:testDebugUnitTest'.
> There were failing tests. See the report at: file:///home/jenkins/jenkins/workspace/java_pr/examples/unitTestExample/build/reports/tests/debug/index.html
```

**Comment by cmelchior** [comment]
Apart from the failing unitTest example this looks good

File: examples/kotlinExample/build.gradle (hunk 1 above)
Comment (donnfelker) [inline examples/kotlinExample/build.gradle]: The kotlin plugin for kotlin adds mavenCentral. If we wnat to remove it, we can try, but it's not what the plugin does. Thoughts?

Comment (cmelchior) [inline examples/kotlinExample/build.gradle]: It seems a bit superfluous, but I don't have any strong opinions on it. So if you think it makes it easier for people to compare their project with our example, just keep it.

**Comment by donnfelker** [comment]
Ok, this has been updated. Upon a successful test pass in Jenkins we should be good. I had to change some tests because I believe I found a bug in Powermock: https://github.com/jayway/powermock/issues/649  The executeTransction() method required a partial mock so that the real method could be called so that the transactions would be executed in the related tests. This caused powermock to return incorrect invocation counts for the method verification in the tests. I've submitted this as a [issue](https://github.com/jayway/powermock/issues/649) on Powermock. To get around this and to verify behavior between the components I created two additional tests that verified the executeTransaction expected behavior.

**Comment by donnfelker** [comment]
@cmelchior can you give this a review when you get a chance?

**Comment by donnfelker** [comment]
@emanuelez @zaki50 Can you guys give this a review since CM is on vacation? :)

File: examples/kotlinExample/build.gradle (hunk 1 above)
Comment (emanuelez) [inline examples/kotlinExample/build.gradle]: jcenter is a superset of maven central so the line is not needed

File: examples/kotlinExample/src/main/kotlin/io/realm/examples/kotlin/KotlinExampleActivity.kt [hunk 2]
```diff
@@ -98,27 +98,24 @@ public class KotlinExampleActivity : Activity() {
...
         // Update person in a transaction
-        realm.beginTransaction()
-        person.name = "Senior Person"
-        person.age = 99
-        showStatus(person.name + " got older: " + person.age)
-        realm.commitTransaction()
+        realm.executeTransaction {
+            person.name = "Senior Person"
+            person.age = 99
+            showStatus(person.name + " got older: " + person.age)
```
Comment (emanuelez) [inline examples/kotlinExample/src/main/kotlin/io/realm/examples/kotlin/KotlinExampleActivity.kt]: I guess this could go outside the transaction

**Comment by emanuelez** [comment]
This looks good apart from a couple of minor comments. I guess it would be interesting to convert one of the examples to use retrolambda, just to show how nice those transactions can look like and to make sure we're compatible with the library. But it's something that could be (and probably should be) done in another PR.

**Comment by cmelchior** [comment]
Beside @emanuelez comments I think this looks good :+1:

File: examples/introExample/src/main/java/io/realm/examples/intro/IntroExampleActivity.java [hunk 3]
```diff
@@ -92,32 +92,40 @@ private void basicCRUD(Realm realm) {
...
+        });
 
         // Delete all persons
-        realm.beginTransaction();
-        realm.allObjects(Person.class).clear();
-        realm.commitTransaction();
+        realm.executeTransaction(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.allObjects(Person.class).clear();
```
Comment (beeender) [inline examples/introExample/src/main/java/io/realm/examples/intro/IntroExampleActivity.java]: Use `realm.delete(Person.class)` instead?

File: examples/threadExample/src/main/java/io/realm/examples/threads/ThreadFragment.java [hunk 4]
```diff
@@ -76,18 +76,24 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
                 return true;
 
             case R.id.action_clear:
-                realm.beginTransaction();
-                realm.clear(Dot.class);
-                realm.commitTransaction();
+                realm.executeTransaction(new Realm.Transaction() {
+                    @Override
+                    public void execute(Realm realm) {
+                        realm.clear(Dot.class);
```
Comment (beeender) [inline examples/threadExample/src/main/java/io/realm/examples/threads/ThreadFragment.java]: `clear` is deprecated now. Use `realm.delete(Dot.class)`.

File: examples/unitTestExample/src/main/java/io/realm/examples/unittesting/ExampleActivity.java [hunk 5]
```diff
@@ -85,9 +85,12 @@ public void onClick(View v) {
 
     private void cleanUp() {
         // Delete all persons
-        realm.beginTransaction();
-        realm.allObjects(Person.class).clear();
-        realm.commitTransaction();
+        realm.executeTransaction(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.allObjects(Person.class).clear();
```
Comment (beeender) [inline examples/unitTestExample/src/main/java/io/realm/examples/unittesting/ExampleActivity.java]: `realm.delete(Person.class)`.

**Comment by beeender** [comment]
Beside some invocation of deprecated methods, :+1:

**Comment by emanuelez** [comment]
retest this please

**Comment by kneth** [comment]
:+1:

**Comment by kneth** [comment]
As #2800 has been merged, I'll close this one.

**Pull request closed without merging**